### DISCUSSION TITLE
In a website-project, how can I additionally render specific dirs/files to additional output formats when calling `quarto render`?

### DISCUSSION THREAD

**Discussion by Gewerd-Strauss** [quarto-dev/quarto-cli#8394] [Q&A]
### Description

Hello,

assume I have a website-project with principal structure:

```txt
report/
    manuscript.qmd
preproject/
    manuscript.qmd
documentation/
    meta/
        docs.qmd
        package_citations.qmd
    methods/
        pls.md
        pca.qmd
    ....
...
```

Files and structure are defined for the website in a root-level `_quarto.yaml`-file. Format-Options are defined in root-level `_quarto-{format}.yaml`-files, e.g. `_quarto-html.yaml`. Execution is kicked off by running `quarto render --to html --to pdf --profile default,html ,pdf` in a terminal.

I now want to render all files documents to html. However, files in subfolders `report` and `preproject` should also be converted to pdf. Documents in other places should only get compiled to HTML. Obviously I cannot just chuck the pdf-yaml file at the project-root, that should result in all files being rendered as pdf as well. 

Instead, you can add a [directory-specific metadata-file](https://quarto.org/docs/projects/quarto-projects.html#directory-metadata), however this seems to have no effect on what files are converted to pdf when running `quarto render --to html --to pdf --profile default,html ,pdf`. Additionally it seems that I am unable to disable pdf-rendering at will unless I remove the directory-level pdf options again.

Am I just missunderstanding how this should be done? 

Thank you.
Sincerely,
~Gw

---

Edit 2024-01-24 11:17: 

`quarto check` returns:

```PS

[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.3.450
      Path: D:\CLI\quarto\bin
      CodePage: 1252

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.11.6
      Path: C:/Users/Claudius Main/AppData/Local/Programs/Python/Python311/python.exe
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with py -m pip install jupyter

[>] Checking R installation...........OK
      Version: 4.3.2
      Path: D:/CLI/r/R-4.3.2
      LibPaths:
        - D:/CLI/r/R-4.3.2/library
      knitr: 1.45
      rmarkdown: 2.25

[>] Checking Knitr engine render......OK
```

**Comment by mcanouil** [comment]
Could you provide a small reproducible example of your setup? (Way more easier to work with a concrete example than to "assume" 😉)

**Reply by Gewerd-Strauss** [reply]
I can do so a bit later today, I am in the process of leaving - but yes.

**Reply by Gewerd-Strauss** [reply]
I've just provided a reprex in [the other thread](https://github.com/quarto-dev/quarto-cli/discussions/8394#discussioncomment-8225518).

**Comment by cderv** [comment]
> Format-Options are defined in root-level _quarto-{format}.yaml-files, e.g. _quarto-html.yaml.

That is interesting. What do you feel you need to use profile to render to a format ? Is this because you are switching for a website project to a default project and you want every .qmd to be a pdf file ? 

> Instead, you can add a [directory-specific metadata-file](https://quarto.org/docs/projects/quarto-projects.html#directory-metadata), however this seems to have no effect on what files are converted to pdf when running quarto render --to html --to pdf --profile default,html ,pdf.

Did you try without resorting to your profile usage for format ? Look at metadata merging (https://quarto.org/docs/projects/quarto-projects.html#metadata-merging) to understand how format key in `_quarto.yml` will be merged with others. 

Maybe we have an issue in merging (and your reproducible example will help), but if you need to have some subfolder to have a specific format, the shared `_metadata.yml` should be the way.

**Reply by Gewerd-Strauss** [reply]
The rationale here is 

1. that I need two files with pdf- and html-output, but not always. When working and just checking via render, I don't want to compile both formats. There are some fairly time-consuming steps associated with each render; some of which lie outsides quarto's control. As such, I would prefer not to have both formats active all the time. To my understanding, the solution for this would be to separate them into profiles. 
2. And all other notes are only part of the website-output - they shouldn't be converted anyways.

While planned a while ago, I am just now trying to set this up, and thus I only now really realised that there's an obvious flaw here: A profile is engaged for all files in the project, not just for some. For this, you would have to set up the pdf-defining `yml`-file only within directories which should _also_ get converted to pdf. My understanding breaks down when it comes to combining the two:

1. project-wide html output
2. conditional, cli-toggleable pdf-output for selected directories

---

Semi-related, but I am currently unable to render to html & pdf simultaneously, because if I am trying to render both simultaneously, e.g. `quarto render documentation --to html, pdf --profile default, pdf, html` (to only render documentation-subdir, so that I don't have to wait for some rather timeconsuming other sections inbetween retries), I get the same consistent error:

```PS
PS D:\Dokumente neu\Obsidian NoteTaking\BE31-Thesis-quarto> quarto render documentation --to html, pdf --profile default, pdf, html

[unrelated pre-render R-script hidden for clarity. It runs to succession.]

[1/5] documentation\meta\docs.qmd
pandoc: pdf: withBinaryFile: does not exist (No such file or directory)
PS D:\Dokumente neu\Obsidian NoteTaking\BE31-Thesis-quarto> 

```

That is, when ignoring the problem at hand here, I am unable to properly test anyways. Even more extreme, narrowing it down further:

```PS

PS D:\Dokumente neu\Obsidian NoteTaking\BE31-Thesis-quarto> quarto render .\documentation\meta\docs.qmd --to html, pdf

[unrelated pre-render R-script hidden for clarity. It runs to succession.]

pandoc pdf
  to: html
  output-file: docs.html
  standalone: true
  title-prefix: B. Sc. Bioengineering Thesis & Code-Documentation
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png

metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  bibliography:
    - 'D:/Dokumente neu/Zotero_Papers/Zotero_papers_autosync.bib'
    - ../../.pandoc/packages.bib
  page-layout: article
  creation date: '2024-01-05 09:55'
  modification date: '2024-01-10 11:07'
  title: docs

pandoc: pdf: withBinaryFile: does not exist (No such file or directory)
PS D:\Dokumente neu\Obsidian NoteTaking\BE31-Thesis-quarto>
```

This occurs if I have a pdf-yaml defined, as well as if I don't have one defined. I could even undefine all formats, not feed a single yaml config. Same issue. Each format works, but only when compiled-to alone. 

As to why this is occuring, I have exhausted seemingly everything I can find on that error, particularly because it doesn't even log _what_ resource is missing. Which makes solving this issue kinda impossible for me. I am not experienced in debugging pandoc; but for me this reads as if pandoc is trying to covert _from_ pdf _to_ html - which will obviously fail.

If that is the case however, I am really confused as to the command-format required for multi-format rendering; as the above conforms to the [documented example](https://quarto.org/docs/get-started/authoring/text-editor.html#rendering).

Executing `quarto render .\documentation\meta\docs.qmd --to html  --to pdf` will render to whatever format is defined last. 

All in all, this is _very_ odd. 

---


The reprex will have to wait till tomorrow. I intended to do this today, but holy shit was the day crappy.

**Reply by Gewerd-Strauss** [reply]
Well, here is a rudimentary setup.

More information is given in `root\selective_output.qmd`, I hope this is sufficient. I'm happy to elaborate, should things still be unclear.

[selective_output.zip](https://github.com/quarto-dev/quarto-cli/files/14029648/selective_output.zip)

**Reply by cderv** [reply]
Thanks for the feedback. Really interesting! This raises indeed the question of workflow for partial rendering (one format and not all ) 🤔 

>  `quarto render documentation --to html, pdf`

about this, try with no space 
````markdown
quarto render documentation --to html,pdf
````

It seems we don't protect against this, and we see ` pdf` format instead of `pdf` - the former is not an existing format if you don't trim the space 😓

**Reply by Gewerd-Strauss** [reply]
> about this, try with no space
> 
> ```
> quarto render documentation --to html,pdf
> ```

As mentioned in the main qmd-file of the reprex, this does not work:

```PS
PS >  quarto render docs --to html,pdf
A new release of Deno is available: 1.28.2 → 1.39.4 Run `deno upgrade` to install it.
[1/2] docs\meta\docs.qmd
ERROR: Unknown format html,pdf

Error: Unknown format html,pdf
    at resolveFormats (file:///D:/CLI/quarto/bin/quarto.js:83183:19)
    at async renderContexts (file:///D:/CLI/quarto/bin/quarto.js:82637:21)
    at async renderFiles (file:///D:/CLI/quarto/bin/quarto.js:86520:28)
    at async renderProject (file:///D:/CLI/quarto/bin/quarto.js:86800:25)
    at async Command.fn (file:///D:/CLI/quarto/bin/quarto.js:90856:32)
    at async Command.execute (file:///D:/CLI/quarto/bin/quarto.js:8437:13)
    at async quarto (file:///D:/CLI/quarto/bin/quarto.js:127545:5)
    at async file:///D:/CLI/quarto/bin/quarto.js:127563:9
```

**Reply by cderv** [reply]
Which version of Quarto are you using ?

**Reply by Gewerd-Strauss** [reply]
uuhm, one sec. will do so, then retest it all - I usually don't run prereleases. Cuz I am not sure, this would entail uninstalling v1.3.450 from PATH first, then cloning down the current master/main and putting it on PATH, correct?

**Reply by cderv** [reply]
Ok I can reproduce. Let me open an issue. I wasn't expecting 🤔 

Either this is an issue, or we just need to provide clearer error. THanks !

**Reply by Gewerd-Strauss** [reply]
So; you still want me to upgrade rn?

**Reply by cderv** [reply]
> Cuz I am not sure, this would entail uninstalling v1.3.450 from PATH first, then cloning down the current master/main and putting it on PATH, correct?

It depends how you install pre-release. It could either replace your existing one, or just install elsewhere and replace PATH. 
Or it could also not be in PATH, and in another directory you would call with absolute path. 

On windows I am using Scoop package manager with https://github.com/cderv/r-bucket#quarto-cli so that I can have both prerelease and release accessible. QVM project (https://github.com/dpastoor/qvm) is also helpful to test some pre-release version. 

Anyhow, no need, I can reproduce with current pre-release

**Reply by cderv** [reply]
Ok this is identified already at 
* https://github.com/quarto-dev/quarto-cli/issues/5947

**Reply by Gewerd-Strauss** [reply]
All right, then I will remain on current stable for now - I'm fairly tight on time and don't want to risk breaking other stuff for now. I'll likely take a look and change once this is fixed probably.

---

I will take a look at the links you've provided; but odd question: Couldn't I just pull the dev version via git and run the config-script, as outlined in the README? 

Seems simpler to me, but then again I'm not experienced with stuff that actually needs compiling, so idk if it works differently here  ¯\\\_(ツ)_/¯.

**Reply by cderv** [reply]
> Couldn't I just pull the dev version via git and run the config-script, as outlined in the README?

Oh you could yes. You'll get the version on main in that case, so could be a bit different than the pre-release build available in Github release. (more unstable as main could be broken more often that pre-release). You'll just need to run configure script to build the binary. It will also put the dev version in PATH, so you'll need to switch PATH if you want back the release one. And updating will require `git pull`

Personally, I find better control in testing pre-release version to use the build installer that going with `git checkout` in the tree, and running configuration to build binary. downloading the pre-release build seems easier to me 😄 

But this is just about workflow preference. So using dev version is also fine!

**Reply by Gewerd-Strauss** [reply]
So the takeaway is that as of right now, this is not doable at-all for now, and for the forseeable future - if I read the other issue you've linked to properly? 
No clue when v1.5. is started/this is getting addressed (and I want to make clear that I am _not_ assuming you drop everything to fix this for me lol.), so if we require 2+ format outputs we just got to suck up and render them all project-wide. Compile all files to all formats.

I'd probably just call via absolute link; I do so on a bunch of other software I run in combination with quarto as well, as I need to have stable and modded releases for various reasons. So git it is probably.

**Reply by cderv** [reply]
The `--to html,pdf` problem is an issue we'll fix in 1.5 (1.4 is closed for now as in release candidate state - we are really picky on what to fix) 

Regarding the workflow to render only a subset of documents in only one format in the project, format which would be not shared throughout the whole project: This seems like a feature request. This is a workflow we may not have considered yet, and we probably need to do some ground based change so that this is possible. 

One thing you can do is open a new clean feature request that would sum up and share your workflow. Explaining clearly and easily so that it can be considered in the future. 

Currently, project profile and maybe [render targets](https://quarto.org/docs/projects/quarto-projects.html#render-targets), coupled with project organisation can help. 

but it seems the workflow you are using, conflicts with some Quarto project limitations & behavior.

**Reply by Gewerd-Strauss** [reply]
> workflow to render only a subset of documents in only one format in the project, format which would be not shared throughout the whole project:

I realise this is effectively the same as saying "render project-wide to format X and a specific subset additionally to format Y", but I never thought about it like that. 

---

I'll then be prepping a hopefully as clean-as-possible FR over the coming days. Thank you.

**Reply by Gewerd-Strauss** [reply]
I somehow missed linking to it here; so just for completeness' sake: 

https://github.com/quarto-dev/quarto-cli/discussions/8455
